### ISSUE TITLE
trailing_stop value relative to profit gains

### ISSUE THREAD

**Issue by yaotian** [freqtrade/freqtrade#1193]
Current situation:
"trailing_stop_positive":  0.01,
"trailing_stop_positive_offset":  0.05,
if profit reach 10%, trailong stop will be enabled. 
if profit reduced 5%, it call sell.

Negative:
It can't catch big up trend like (go up 5%, go down 2%, go up 5%, down down 2%)

=============================================

New Suggestion:
trailing_stop enabler: 10%
profit reduce checker: 50%

Example:
buy on 100
case 1:price rise 10% to 110, if profit (10%) *reduced*  50%( i.e. 10% * 50% = 5% ), it sell
case2: price rise 10% and didn't go back. Continue to 20%. If profit(20%) *reduce* 50% ( i.e 20% * 50% = 10%), it sell

Explain:
More profit more trailing_stop. 
Case 1 stop point is 5%. 
Case 2 stop point is 10%
It helps to catch big up trend

**Comment by xmatthias** [comment]
Hi,
sorry for the late response, i needed some time to think about this.

if i understand this correctly - you would want a parameter which only acts in positive range (once profit_offset is reached), and is based on %profit - not overall percent?

This would make sense and would be a great addition.

**Comment by yaotian** [comment]
@xmatthias 

Logic is like this.

```
        if 'profit_stop_enable_pct' in self.config and 'profit_stop_pct' in self.config:
            trade.adjust_stop_loss(current_rate, self.stoploss)

            profit_stop_enable_pct = self.config.get('profit_stop_enable_pct', 0.0) 
            profit_stop_pct = self.config.get('profit_stop_pct', 0.0) 
            max_profit = trade.calc_profit_percent(trade.max_rate)
            if max_profit > profit_stop_enable_pct:
                stoploss = max_profit * (1 - profit_stop_pct)
                trade.adjust_stop_loss(current_rate, stoploss)
 
```

**Comment by ynzheng** [comment]
"stoploss": -0.05,
    "trailing_stop": true,
    "trailing_stop_positive": 0.05,
    "trailing_stop_positive_offset": 0.10,
    "minimal_roi": {
        "0":  0.20
    },

你举的例子，这样就可以把

**Comment by mrsegen** [comment]
I'm considering working on this, but want to check in first regarding how the circumstances have shifted since this was posted in 2018.

The configuration as @ynzheng mentioned is now possible (i.e. [trailing stoploss once offset is reached](https://www.freqtrade.io/en/latest/stoploss/#trailing-stop-loss-only-once-the-trade-has-reached-a-certain-offset)). It seems that wasn't available at the this issue was written (no mention in the [2019.6 docs](https://www.freqtrade.io/en/2019.6/stoploss/)).

But trailing offset by _offset_ doesn't deliver what is requested here: trailing offset by _proportional_ difference. If I wind up picking this up, I'd likely start with figuring out how to add test cases similar to `test_tsl_only_offset_reached` within test_freqtradebot.py, and move into the code from there. I imagine this would likely involve some maintenance of the interface for the `stoploss_on_exchange` functionality.

**Comment by xmatthias** [comment]
> I imagine this would likely involve some maintenance of the interface for the stoploss_on_exchange functionality.

not necessarily - if the stoploss is on or off the exchange should not really matter on HOW the stoploss value itself is calculated. The only thing is - a trailing stoploss should never be lowered once it's been at a certain level.

I understand the main points behind this issue (the more profit you make, the more room you give the trade to temporarily pull back).

I am however not yet certain on the way to do this. One way is by introducing a setting (e.g. `trailing_stop_positive_ratio`) - and have this calculation pre-made (as requested above).

A most likely better (more flexible) approach would be to allow the user to define a custom stoploss directly from the strategy (which could solve #3206 and #3664)
this would also allow 

* "non-trailing" custom stoplosses (e.g. when positive by 20%, set stoploss to 10% to avoid going to negative again) - but without trailing in general.
* custom stoploss per pair (give BTC/USDT more room than XRP/USDT - maybe because we know that XRP/USDT is more likely to fail on our strategy ...)
* base your stoploss on the pair's latest performance 
* ... (limited by your imagination and other market participants moving the price in the right direction)


Having this functionality would turn this PR into a documentation problem, where we could provide an example on how to do this within the `get_stoploss_price()` callback (which would leave a lot of flexibility to the user ... as maybe i want it trailing one way up to a certain point, and trailing another way after that).

Obviously, that would be a "complexer" approach - but we'll go that route anyway - and that's mainly the point why this issue has been left lying around for so long (we will try to allow custom pricing for all orders at a certain point). 

Adding another way to calculate trailing stoploss before that would most likely complicate things quite a bit ...

**Comment by mrsegen** [comment]
@xmatthias If I'm understanding your suggestion, it's to _not_ implement proportional trailing stoploss via additional config variables. But, rather, to enable arbitrary stoploss functionality via custom strategies (which would of course implement the IStrategy interface).

Given that settings can either set within a strategy and/or overridden in a config.json file, I yet follow following what that means concretely. (I am still early on in writing custom strategies). 

Any approach to this would likely involve a change to the strategy interface (and new strategy-processing functionality).

**Comment by xmatthias** [comment]
It'd be a change to the strategy interface, yes - but a rather "Minor" one, and one that can be made non-breaking (just return self.stoploss in the default method that is supposed to be overwritten).

Maybe have a look at #3886 for an example on how this could look like (although for stoploss, it's a bit more complex)...

Alternatives are in the [advanced strategy](https://www.freqtrade.io/en/latest/strategy-advanced/#trade-entry-buy-order-confirmation) documentation.

Don't get me wrong - this WILL require some deep knowledge on bot functionings - i've got a rough idea, but it's pretty improtant to catch all occurances of stoploss.

I will try to work on this after my current enhancement - which i HOPE to complete today or tomorrow ...

**Comment by xmatthias** [comment]
As promised, #4089 turns this issue into a documentation point. 
Theres an example in the documentation build specifically for this issue.

**Comment by mrsegen** [comment]
@xmatthias I haven't _yet_ figured out how to use the changed code to implement the _proportional_ trailing stop loss. Still reviewing #4089 though.
